src/libstd/path.rs

```diff
@@ -0,0 +1,2415 @@
+                unsafe { os_str_as_u8_slice(s).len() }
+            }
+            match *self {
+                Verbatim(x) => 4 + os_str_len(x),
+                VerbatimUNC(x,y) => 8 + os_str_len(x) +
+                    if os_str_len(y) > 0 { 1 + os_str_len(y) }
+                    else { 0 },
+                VerbatimDisk(_) => 6,
+                UNC(x,y) => 2 + os_str_len(x) +
+                    if os_str_len(y) > 0 { 1 + os_str_len(y) }
+                    else { 0 },
+                DeviceNS(x) => 4 + os_str_len(x),
+                Disk(_) => 2
+            }
+
+        }
+
+        #[inline]
+        pub fn is_verbatim(&self) -> bool {
+            use self::Prefix::*;
+            match *self {
+                Verbatim(_) | VerbatimDisk(_) | VerbatimUNC(_, _) => true,
+                _ => false
+            }
+        }
+
+        #[inline]
+        pub fn is_drive(&self) -> bool {
+            match *self {
+                Prefix::Disk(_) => true,
+                _ => false,
+            }
+        }
+
+        #[inline]
+        pub fn has_implicit_root(&self) -> bool {
+            !self.is_drive()
+        }
+    }
+
+    pub const MAIN_SEP: u8 = b'\\';
+    pub const MAIN_SEP_STR: &'static str = "\\";
+}
+
+////////////////////////////////////////////////////////////////////////////////
+// Misc helpers
+////////////////////////////////////////////////////////////////////////////////
+
+// Iterate through `iter` while it matches `prefix`; return `None` if `prefix`
+// is not a prefix of `iter`, otherwise return `Some(iter_after_prefix)` giving
+// `iter` after having exhausted `prefix`.
+fn iter_after<A, I, J>(mut iter: I, mut prefix: J) -> Option<I> where
+    I: Iterator<Item=A> + Clone, J: Iterator<Item=A>, A: PartialEq
+{
+    loop {
+        let mut iter_next = iter.clone();
+        match (iter_next.next(), prefix.next()) {
+            (Some(x), Some(y)) => {
+                if x != y { return None }
+            }
+            (Some(_), None) => return Some(iter),
+            (None, None) => return Some(iter),
+            (None, Some(_)) => return None,
+        }
+        iter = iter_next;
+    }
+}
+
+// See note at the top of this module to understand why these are used:
+unsafe fn os_str_as_u8_slice(s: &OsStr) -> &[u8] {
+    mem::transmute(s)
+}
+unsafe fn u8_slice_as_os_str(s: &[u8]) -> &OsStr {
+    mem::transmute(s)
+}
+
+////////////////////////////////////////////////////////////////////////////////
+// Cross-platform parsing
+////////////////////////////////////////////////////////////////////////////////
+
+/// Says whether the path ends in a separator character and therefore needs to
+/// be treated as if it ended with an additional `.`
+fn has_suffix(s: &[u8], prefix: Option<Prefix>) -> bool {
+    let (prefix_len, verbatim) = if let Some(p) = prefix {
+        (p.len(), p.is_verbatim())
+    } else { (0, false) };
+    if prefix_len > 0 && prefix_len == s.len() && !verbatim { return true; }
```

Comment:
I will probably have many questions like this over time, so some of them can be batted away, but what do prefixes and verbatim things have to do with whether a path has a trailing `/`? Could you add a comment with an example of a path that this is returning `true` for? (but wouldn't return `true` below)